pandas/io/sql.py

```diff
@@ -1364,7 +1364,7 @@ class PandasSQL(PandasObject, ABC):
     Subclasses Should define read_query and to_sql.
     """
 
-    def __enter__(self):
+    def __enter__(self) -> PandasSQL:
```

Comment:
This returns the relevant subclass (`SQLDatabase` & `SQLDatabase`) when those are used. Add a `_PandasSqlT = TypeVar("_PandasSqlT", bound="PandaSQL")` and use that?

I've made some arguments that we should allow use of `typing_extensions.Self` in some other threads, but that's a bit of a different issue.

Reply:
I would love using Self here